File: frontend/src/app/modules/todo/services/TodoService.ts
import {Injectable} from '@angular/core';
import {Todo} from '../model/Todo';
import {HttpClient} from '@angular/common/http';
import {Observable} from 'rxjs';
import {environment} from '../../../../environments/environment';
import {EventDialogComponent} from '../../public/component/dialog/eventDialog.component';
import {MatDialog} from '@angular/material';

@Injectable({
    providedIn: 'root',
})
export class TodoService {

    todoToUpdate: Todo;

    constructor(private http: HttpClient, private dialog: MatDialog) {
    }

    /**
     * Returns a list of {Todo[]} todo
     */
    getTodoList(): Observable<Todo[]> {
        return this.http.get<Todo[]>(environment.apiUrl + '/getAllTodos');
    }

    /**
     * Add Todo
     * @param {Todo} param data
     * @return {Observable<Todo>} Returns an Observable of todo
     */
    addTodo(data: Todo): Observable<Todo> {
        return this.http.post<Todo>(environment.apiUrl + '/addTodo', data);
    }

    /**
     * Record an item that is to be updated
     * @param {Todo} param data
     */
    setItemToUpdate(data: Todo): void {
        this.todoToUpdate = data;
    }

    /**
     * Returns an item that is to be updated
     * @return {Todo} Returns an item that is to be updated
     */
    getItemToUpdate(): Todo {
        return this.todoToUpdate;
    }

    /**
     *  Delete a Todo item
     *  @param {string} param id
     *  @return {Observable<Todo>} Returns an Observable of todo
     */
    deleteTodo(id: string): Observable<Todo> {
        return this.http.delete(environment.apiUrl + '/deleteTodo/' + id);
    }

    /**
     *  Update Todo Item
     *  @param {Todo} param data
     *  @return {Observable<Todo>} Returns an Observable of todo
     */
    updateTodo(data: Todo): Observable<Todo> {
        return this.http.put(environment.apiUrl + '/updateTodo/', data);
    }

    /**
     *  Delete Multiple Todos
     *  @param {Todo[]} param idList
     *  @return {Observable<Todo>} Returns an Observable of todo
     */
    deleteMultipleTodos(idList: Todo[]): Observable<Todo> {
        return this.http.delete(environment.apiUrl + '/deleteMultipleTodos/' + idList);
    }

    /**
     * Dialog Box to display messages
     * @param {string} pram dialogMessage
     */
    messageDialogBox(dialogMessage: string) {
        this.dialog.open(EventDialogComponent, {
            data: {
                message: dialogMessage
            }
        });
    }
}

File: backend/src/main/java/com/todo/todoapp/service/TodoService.java
package com.todo.todoapp.service;

import com.todo.todoapp.dto.TodoDTO;
import com.todo.todoapp.model.Todo;

import java.util.List;

public interface TodoService {

    /**
     * @param todoDTO : todoDTO
     * @implNote Method is responsible to save  request into DB
     */
    Todo addTodo(final TodoDTO todoDTO);

    /**
     * @return List of todoName
     * @implNote Method is responsible to get  todos
     */
    List<Todo> getTodos(final String todoName);

    /**
     * @implNote Method is responsible to get all todos
     */
    List<Todo> getAllTodos();
    /**
     * @param id : id
     * @implNote Method is responsible to delete from DB
     */
    void deleteTodo(final Long id);
    /**
     * @param todoIds : todoIds
     * @implNote Method is responsible to delete multiple records from DB
     */
    void deleteMultipleTodos(final List<Long> todoIds);

    /**
     * @param  todoDTO : todoDTO
     * @implNote Method is responsible to update records.
     */
    Todo updateTodo(final TodoDTO todoDTO);
}
